File: examples/python/07_sessions.py
import time

import audio_sdk


def main() -> None:
    manager = audio_sdk.AudioManager()

    def on_event(event: object) -> None:
        print(
            f"event={audio_sdk.event_type_name(event.type)} "
            f"device={event.device_id} "
            f'details="{event.details}"'
        )

    manager.set_event_callback(on_event)
    input_device = audio_sdk.select_first_device(manager, audio_sdk.DeviceDirection.INPUT)
    output_device = audio_sdk.select_first_device(manager, audio_sdk.DeviceDirection.OUTPUT)
    print(f"Selected input device: {input_device.name} ({input_device.id})")
    print(f"Selected output device: {output_device.name} ({output_device.id})")

    path = audio_sdk.recording_path("python_session_example.wav")

    recording_session = manager.create_recording_session(str(path))
    audio_sdk.require(recording_session.start(), "start recording session")
    print(f"Recording session active={recording_session.active()}")
    audio_sdk.sleep_for_recording()
    audio_sdk.require(recording_session.stop(), "stop recording session")

    playback_session = manager.create_playback_session(str(path))
    audio_sdk.require(playback_session.start(), "start playback session")
    while playback_session.active():
        time.sleep(0.1)

    print(f"Session-based record/playback completed at {path}")


if __name__ == "__main__":
    main()

File: examples/cpp/07_sessions.cpp
#include <chrono>
#include <iostream>
#include <thread>

#include "example_utils.h"

int main() {
    audio_sdk::AudioManager manager;

    manager.SetEventCallback([](const audio_sdk::Event& event) {
        std::cout << "event=" << audio_sdk_examples::EventTypeName(event.type)
                  << " device=" << event.device_id
                  << " details=\"" << event.details << "\"\n";
    });

    if (!audio_sdk_examples::SelectFirstDevice(manager, audio_sdk::DeviceDirection::kInput)) {
        return 1;
    }
    if (!audio_sdk_examples::SelectFirstDevice(manager, audio_sdk::DeviceDirection::kOutput)) {
        return 1;
    }

    const std::filesystem::path path = audio_sdk_examples::RecordingPath("cpp_session_example.wav");

    auto recording_session = manager.CreateRecordingSession(path.string());
    if (!recording_session || !audio_sdk_examples::Require(recording_session->Start(), "start recording session")) {
        return 1;
    }

    std::cout << "Recording session active=" << (recording_session->active() ? "yes" : "no") << '\n';
    audio_sdk_examples::SleepForRecording();

    if (!audio_sdk_examples::Require(recording_session->Stop(), "stop recording session")) {
        return 1;
    }

    auto playback_session = manager.CreatePlaybackSession(path.string());
    if (!playback_session || !audio_sdk_examples::Require(playback_session->Start(), "start playback session")) {
        return 1;
    }

    while (playback_session->active()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    std::cout << "Session-based record/playback completed at " << path << '\n';
    return 0;
}
